File: src/main/java/compiler/block/structures/objects/ObjectBlock.java
package compiler.block.structures.objects;

import compiler.block.Block;
import compiler.block.imports.ImportBlock;
import compiler.block.structures.FileBlock;
import compiler.tokenizer.Token;

// Creation of a new object and storing to a variable
public class ObjectBlock extends Block {

    Token className;
    Token variableName;
    Token operator;
    Token newKeyword;
    Token initClassName;

    String directory = "";

    public ObjectBlock(Block superBlock, Token className, Token variableName , Token operator, Token newKeyword , Token initClassName) {
        super(superBlock, false, true);
        this.className = className;
        this.variableName = variableName;
        this.operator = operator;
        this.newKeyword = newKeyword;
        this.initClassName = initClassName;
    }

    @Override
    public void init() {
        directory = getDirectory();
    }

    // Gets the directory of the class using the Imports. Otherwise assumes class is  in the same package
    public String getDirectory(){
        // Get the FileBlock to find the imports
        Block block = this;
        while(!(block instanceof FileBlock)){
            block = block.getSuperBlock();
        }

        // Get the directory of the Object
        for(Block sub : block.getSubBlocks()){
            if(sub instanceof ImportBlock && ((ImportBlock)sub).fileName.equals(className.getToken())){
                return ((ImportBlock)sub).directory;
            }
        }
        return "";
    }

    @Override
    public String getName() {
        return variableName.getToken();
    }

    @Override
    public String getValue() {
        return null;
    }

    @Override
    public String getType() {
        return className.getToken();
    }

    @Override
    public String getOpeningCode() {
        return "";
    }

    @Override
    public String getBodyCode() {
        return "mv.visitTypeInsn(NEW, \""+directory+"/"+className+"\");\n" +
                "mv.visitInsn(DUP);\n" +
                "mv.visitMethodInsn(INVOKESPECIAL, \""+directory+"/"+className+"\", \"<init>\", \"()V\", false);\n" +
                "mv.visitVarInsn(ASTORE,"+ getId() +");\n";
    }

    @Override
    public String getClosingCode() {
        return "";
    }
}

File: src/main/scala/compiler/block/structures/objects/ObjectBlock.scala
package compiler.block.structures.objects

import compiler.{Parameter, Utils}
import compiler.block.Block
import compiler.block.imports.ImportBlock
import compiler.block.packages.PackageBlock
import compiler.block.structures.FileBlock
import compiler.block.structures.classes.ClassBlock
import compiler.symbol_table.SymbolTable

// Creation of a new object and storing to a variable
class ObjectBlock(var superBlockInit: Block, var className: String, var variableName: String, var operator: String, var newKeyword: String, var initClassName: String, var params: Array[Parameter]) extends Block(superBlockInit, false, true) {

  private var parameterString: String = ""
  private var argumentString: String = ""
  private[objects] var directory: String = ""

  def init() {
    if (className == getClassName) {
      directory = getPackage
    }
    else {
      directory = getDirectory
    }

    // Get the type of the parameters
    for (param <- params) {
      param.setType(SymbolTable.getInstance.getValue(Utils.getMethod(this), param.getName).getType)
      parameterString += param.getAsmType
      argumentString += "mv.visitIntInsn(ILOAD, " + SymbolTable.getInstance.getValue(Utils.getMethod(this), param.getName).getId + ");"
      println("Params" + param)
    }
    println(parameterString)
    println(argumentString)
  }

  // Gets the directory of the class using the Imports. Otherwise assumes class is  in the same package
  def getDirectory: String = {
    // Get the FileBlock to find the imports
    var block: Block = this
    while (!(block.isInstanceOf[FileBlock])) {
      {
        block = block.superBlock
      }
    }
    // Get the directory of the Object
    for (sub <- block.subBlocks) {
      if (sub.isInstanceOf[ImportBlock] && (sub.asInstanceOf[ImportBlock]).fileName == className) {
        return (sub.asInstanceOf[ImportBlock]).directory
      }
    }
    return ""
  }

  // Gets the directory of the class using the Imports. Otherwise assumes class is  in the same package
  def getPackage: String = {
    // Get the FileBlock to find the imports
    var block: Block = this
    while (!(block.isInstanceOf[FileBlock])) {
      {
        block = block.superBlock
      }
    }
    // Get the directory of the Object
    for (sub <- block.subBlocks) {
      if (sub.isInstanceOf[PackageBlock]) {
        return (sub.asInstanceOf[PackageBlock]).directory
      }
    }
    return ""
  }

  // Returns the main class name for the file
  def getClassName: String = {
    // Get the FileBlock to find the imports
    var block: Block = this
    while (!(block.isInstanceOf[ClassBlock])) {
      {
        block = block.superBlock
      }
    }
    // Get the directory of the Object
    return block.getName
  }

  def getName: String = variableName

  def getValue: String = null

  def getType: String = className

  def getOpeningCode: String = {
    return "mv.visitTypeInsn(NEW, \"" + directory + (if (directory == "") ""
    else "/") + className + "\");\n" + "mv.visitInsn(DUP);\n" + argumentString + "mv.visitMethodInsn(INVOKESPECIAL, \"" + directory + (if (directory == "") ""
    else "/") + className + "\", \"<init>\", \"(" + parameterString + ")V\", false);\n" + "mv.visitVarInsn(ASTORE," + id + ");\n"
  }

  def getClosingCode: String = {
    return ""
  }

  override def toString: String = "object: " + className + " " + variableName + " = new " + initClassName + "()"

}